### DISCUSSION TITLE
Query composition

### DISCUSSION THREAD

**Discussion by selimb** [prisma/prisma#12047] [Q&A] [answered]
How would you compose/chain query options (i.e. options to `findMany`) from small reusable functions, e.g. filters.
For instance, I would do something like this in Django (see https://simpleisbetterthancomplex.com/tips/2016/08/16/django-tip-11-custom-manager-with-chainable-querysets.html for more details):
```python
Document.objects.pdfs().smaller_than(1000).order_by("name")
```

where `pdfs` and `smaller_than` are simple functions that augment a query:
```python
    def pdfs(self):
        return self.filter(file_type='pdf')

    def smaller_than(self, size):
        return self.filter(size__lt=size)
```
This pattern is also used in [elixir ecto](https://blog.drewolson.org/composable-queries-ecto).

**Comment by selimb** [comment]
Found https://twitter.com/RomanSndlr/status/1479530356902608901?s=20&t=M_g6z-QPU5lnu0O1YeB91A and https://github.com/prisma/prisma/discussions/3929#discussioncomment-98676. So I guess one could do:
```typescript
function whereIsPdf(): Prisma.DocumentWhereInput {
  return {file_type: "pdf"}
}

function whereSizeSmallerThan(size: number): Prisma.DocumentWhereInput {
  return {size: {lt: size}}
}

prisma.document.findMany({
  where: {
    ...whereIsPdf(),
    ...whereSizeSmallerThan(1000)
  },
  orderBy: {name: "asc"}
})
```
Does this look idiomatic?

I suspect this can get more complicated when combining relation filters, e.g. [this example](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#get-all-user-records-where-all-posts-are-published-and-a-least-one-related-post-mentions-prisma)
```typescript
// Original
const result = await prisma.user.findMany({
  where: {
    post: {
      every: {
        published: true
      },
      some: {
        content: {
          contains: "Prisma"
        }
      }
    }
  }
}

function whereAllPostsPublished(): Prisma.UserWhereInput {
  return {post: {every: {published: true}}}
}
function whereOnePostContainsContent(text: string): Prisma.UserWhereInput {
  return {post: {some: {content: {contains: text}}}}
}
await prisma.user.findMany({
  where: {
    ...whereAllPostsPublished(), // Oops, this will be overriden by the object below
    ...whereOnePostContainsContent("Prisma"),
  }
}
```
In this case a deep object merge would be required?

**Comment by austincrim** [comment, marked as answer]
Hey @selimb 👋 ,

You might have some difficulty trying to adopt this pattern with Prisma for some of the reasons you outlined, however since Prisma expect plain objects for query options, any method of generating those objects should work fine. Deep merging objects could get tricky depending on the complexity of your queries.

In general, I have found that it is easier to extract repeated queries rather than specific conditions, but I would be interested to hear if you come up with a pattern you like!

**Comment by selimb** [comment]
Hello @austincrim 

Thanks for your reply!

I see. I'll keep experimenting and hopefully get back to you with more info :)

One particular use case that comes to mind is permission-related filtering, e.g. admins can view all objects but other users can only view objects they own:

```python
def for_user(self, user):
    return self if user.is_admin else self.filter(owner=user)
```

What's nice is that in a typical CRUD API this can be used both to list all objects (`findMany`) and retrieve a single object (`findUnique`). Extracting queries rather than conditions would require duplicating the filter.

**Comment by wodCZ** [comment]
Hello @selimb, sorry to bother you, but have you found a sane way to compose queries?

I'm currently struggling with something similar. I'll try to explain my thinking process, maybe there's a flaw in it, maybe it can already be achieved, or maybe it will just stay here as an example of what can't be achieved.

I have an `AccountsService`, with `findOne` method. It's responsibility is to filter the accounts:
```ts
  findOne(id: string) {
    return this.prismaService.account.findFirst({ where: { id, active: true } });
  }
```

In InvitesService I'm trying to `findOne` the Account using the above (InviteService doesn't need to know the `active: true` business requirement of `AccountsService`, that's why it's querying the `AccountsService` dedicated method).
```ts
  async inviteUser(userId: string, accountId: string, email: string) {
    const account = await this.accountsService.findOne(accountId);
    if (!account) {
      throw new NotFoundException('Account not found');
    }

    // TODO: check if user is already a member of the account?
    // TODO: check if user is already invited to the account?
  }
```

Unfortunately `InviteService` requires `account.managers` and `account.owner` relations to further proceed. This would be easy, I'd just add `include: { owner: true, managers: true }` to join the relevant relations.

But the query itself is hidden within `AccountService.findOne` method, and I certainly don't want to join these relations at all places where I call this method. I could add a method `AccountService.fineOneWithOwnerAndManagers` with duplicated query logic, but it already code-smells.

A better idea seems to be to add a parameter, so the `findOne` method is doing the `where`, and only passing through the rest of the options:
```ts
  findOne(id: string, options?: Omit<AccountFindFirstArgs, 'where'>) {
    return this.prismaService.account.findFirst({
      where: { id, active: true },
      ...options,
    });
  }
```

Unfortunately, this doesn't really work, the type isn't affected by the passed options:
```ts
    const account = await this.accountsService.findOne(accountId, {include: { owner: true, managers: true}});
    if (!account) {
      throw new NotFoundException('Account not found');
    }
    const owner = account.owner; // <---- TS2339: Property 'owner' does not exist on type 'Account'.
```

Maybe this is simply a TypeScript limitation, that it doesn't *propagate* the options through the function call. I've tried appending both `as const` and `Prisma.validate` to the `include` argument in the last snipped, none of them changed anything.

I'd appreciate any idea or comment. Thanks 🙏

**Reply by selimb** [reply]
@wodCZ I could be wrong, but I don't believe Typescript's type inference is able to automatically infer these kinds of "conditional" return types, i.e. return types where the shape is influenced by the parameters.
What I've had to do is rewrite (some of) the generic parts of Prisma's types.
```typescript
import type * as P from "@prisma/client";

export type AccountIncludeOptions = {
  includeOwner?: boolean;
  includeManagers?: boolean;
}

export type Account<I extends AccountIncludeOptions = {}> = P.Account & 
  (I["includeOwner"] extends true ? {owner: User} : {}) &  // Replace User with the actual type of `owner` 
  (I["includeManagers"] extends true ? {managers: User[]} : {});  // Replace `User[]` with the actual type of `managers`

findOne<I extends AccountIncludeOptions = {}>(
  id: string, 
  opts?: {include?: I}
): Promise<Account<I> | null> {
  return this.prismaService.account.findFirst(
    where: { id, active: true },
    include: {
      owner: opts?.include?.includeOwner ?? false,
      managers: opts?.include?.includeManagers ?? false,
    }
  )
}

// Usage:
findOne(accountId, {include: {includeOwner: true, includeManagers: true}})
```
Is this tedious? Absolutely XD

**Reply by wodCZ** [reply]
Ah, I see, that could work (with both eyes closed anyway 😁 ), thanks for sharing ❤️ 

For reference, I've brought this discussion up at the Prisma Slack and got [this response](https://prisma.slack.com/archives/CCWDULGUW/p1661797573596809?thread_ts=1661788353.205199&cid=CCWDULGUW) by Austin:
> We do not have an official recommendation or instructions on how to create these kind of reusable services. We are always interested in learning more about these kind of use cases, though!
There is a ton of discussion on this and related topics in [this existing feature request](https://github.com/prisma/prisma/issues/5273). It would be great if you could add to the conversation there; it really helps our engineers when we can understand more use cases.

At least now I'm *at peace* knowing that it's not an incorrect approach, it's just not settled on how to approach this.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
